### PR TITLE
[Kotlin][Client] Added Kotlinx Serialization for JVM/Retrofit2/OkHttp3&4

### PR DESCRIPTION
**Pull request by kuFEAR**
Added supporting for Kotlinx.serialization 1.1.0 on JVM with base set of JVM type serialize adapters, also update Kotlin 1.4.30 and Gradle wrapper 6.8.3

### PR checklist

- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] If contributing template-only or documentation-only changes which will change sample output, [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) beforehand.
- [x] Run the shell script `./bin/generate-samples.sh`to update all Petstore samples related to your fix. This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. These must match the expectations made by your contribution. You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/config/java*`. For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`
- [x] Copy the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

@jimschubert (2017/09) ❤️, @dr4ke616 (2018/08) @karismann (2019/03) @Zomzog (2019/04) @andrewemery (2019/10) @4brunu (2019/11) @yutaka0m (2020/03)

### CODE REVIEW

**Comment by wing328** [comment]
@kuFEAR thanks for the PR. Can you please resolve the merge conflicts when you've time?

**Comment by wing328** [comment]
Shippable CI failed with the following errors:
```
===> Downloaded package, caching at /root/.cache/rebar3/hex/hexpm/packages/rebar3_proper-0.12.0.tar
===> Compiling rebar3_proper
===> Verifying dependencies...
===> Fetching jsx v2.9.0
===> Downloaded package, caching at /root/.cache/rebar3/hex/hexpm/packages/jsx-2.9.0.tar
===> Fetching proper v1.3.0
===> Downloaded package, caching at /root/.cache/rebar3/hex/hexpm/packages/proper-1.3.0.tar
===> Compiling proper
===> Compiling jsx
===> Compiling petstore
Downloading https://services.gradle.org/distributions/gradle-6.0-bin.zip
.........10%.........20%.........30%.........40%.........50%.........60%.........70%.........80%.........90%.........100%

Welcome to Gradle 6.0!

Here are the highlights of this release:
 - Substantial improvements in dependency management, including
   - Publishing Gradle Module Metadata in addition to pom.xml
   - Advanced control of transitive versions
   - Support for optional features and dependencies
   - Rules to tweak published metadata
 - Support for Java 13
 - Faster incremental Java and Groovy compilation
 - New Zinc compiler for Scala
 - VS2019 support
 - Support for Gradle Enterprise plugin 3.0

For more details see https://docs.gradle.org/6.0/release-notes.html

Starting a Gradle Daemon (subsequent builds will be faster)

> Configure project :
Kotlin Multiplatform Projects are an experimental feature.

Please wait while Kotlin/Native compiler 1.3.72 is being installed.
Download https://download.jetbrains.com/kotlin/native/builds/releases/1.3.72/linux/kotlin-native-linux-1.3.72.tar.gz (102.85 MB)
Download https://download.jetbrains.com/kotlin/native/builds/releases/1.3.72/linux/kotlin-native-linux-1.3.72.tar.gz finished, took 2 s 784 ms
Unpack Kotlin/Native compiler to /root/.konan/kotlin-native-linux-1.3.72
Unpack Kotlin/Native compiler to /root/.konan/kotlin-native-linux-1.3.72 finished, took 3 s 178 ms

FAILURE: Build failed with an exception.

* Where:
Build file '/root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/build.gradle' line: 31

* What went wrong:
A problem occurred evaluating root project 'kotlin-client-petstore-multiplatform'.
> Operation is not supported for read-only collection

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 29s

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.0/userguide/command_line_interface.html#sec:command_line_warnings
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (bundle-test) on project KotlinMultiPlatformClientTests: Command execution failed.: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
```

Looks like the cause is `/root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/build.gradle` line: 31 ? Any idea why it fails?

**Comment by kuFEAR** [comment]
Oh, I'll fix multiplatform SafeEnum build today

**Comment by wing328** [comment]
Now Shippable CI reports the following errors:
```
> Task :compileKotlinIosArm64 SKIPPED
> Task :compileKotlinIosX64 SKIPPED

> Task :compileKotlinJs FAILED
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (144, 79): Using 'list: KSerializer<List<T>>' is an error. Deprecated in the favour of the same extension from builtins package
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (145, 60): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (188, 79): Using 'list: KSerializer<List<T>>' is an error. Deprecated in the favour of the same extension from builtins package
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/PetApi.kt: (189, 60): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/StoreApi.kt: (107, 98): Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
public fun <T : Any> KClass<???>.serializer(): KSerializer<???> defined in kotlinx.serialization
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/StoreApi.kt: (107, 125): Unresolved reference. None of the following candidates is applicable because of receiver type mismatch: 
public fun <T : Any> KClass<???>.serializer(): KSerializer<???> defined in kotlinx.serialization
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/StoreApi.kt: (108, 60): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt: (106, 81): Using 'list: KSerializer<List<T>>' is an error. Deprecated in the favour of the same extension from builtins package
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt: (107, 60): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt: (147, 81): Using 'list: KSerializer<List<T>>' is an error. Deprecated in the favour of the same extension from builtins package
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/apis/UserApi.kt: (148, 60): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/infrastructure/Base64ByteArray.kt: (10, 52): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/infrastructure/OctetByteArray.kt: (10, 52): Using 'withName(String): SerialDescriptor' is an error. Deprecated in the favour of PrimitiveDescriptor factory function
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/ApiResponse.kt: (18, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/ApiResponse.kt: (19, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/ApiResponse.kt: (20, 47): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Category.kt: (18, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Category.kt: (19, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Category.kt: (20, 47): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (18, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (19, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (20, 47): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (46, 26): Type mismatch: inferred type is KClass<Order.Status.Serializer> but KClass<out KSerializer<*>> was expected
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (47, 23): 'serialName' overrides nothing
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (47, 65): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Order.kt: (52, 29): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (20, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (21, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (22, 47): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (35, 34): Unresolved reference: Required
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (36, 39): Unresolved reference: Required
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (48, 26): Type mismatch: inferred type is KClass<Pet.Status.Serializer> but KClass<out KSerializer<*>> was expected
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (49, 23): 'serialName' overrides nothing
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (49, 65): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Pet.kt: (54, 29): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Tag.kt: (18, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Tag.kt: (19, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/Tag.kt: (20, 47): Unresolved reference: SafeEnumAdapter
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/User.kt: (18, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/User.kt: (19, 47): Unresolved reference: SafeEnum
e: /root/src/github.com/OpenAPITools/openapi-generator/samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/models/User.kt: (20, 47): Unresolved reference: SafeEnumAdapter

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileKotlinJs'.
> Compilation error. See log for more details

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.0/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 46s
1 actionable task: 1 executed
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:1.2.1:exec (bundle-test) on project KotlinMultiPlatformClientTests: Command execution failed.: Process exited with an error: 1 (Exit value: 1) -> [Help 1]
```

When you've time, please take another look. If you need help from us, let us know.

**Comment by kuFEAR** [comment]
@wing328  I'll wait kotlin update from this PR https://github.com/OpenAPITools/openapi-generator/pull/7149/files, now we have some overlaps in changes and double work

**Comment by kuFEAR** [comment]
@wing328 Also I want to add exact SerializeAdapter instead @ContextualSerializer for each custom properties, how I can extend CodegenProperty for that, and is it acceptable changes for generators in common? Now I can use only isPrimitiveType/isEnum/complexType and some times it works incorrect
the second moment, that we add import section with types which not present in generated models and then they highlighted as unused

**Comment by kuFEAR** [comment]
Hi @wing328 , why new gradle is not apply on CI? I see `Welcome to Gradle 4.9!` in logs, but should be 6.7

**Comment by kuFEAR** [comment]
@4brunu @wing328 Hi, I fixed the builds on ci, can you review PR?

File: modules/openapi-generator/src/main/resources/kotlin-client/build.gradle.mustache [hunk 1]
```diff
@@ -26,13 +26,17 @@ buildscript {
     }
     dependencies {
         classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+        classpath "org.jetbrains.kotlin:kotlin-serialization:$kotlin_version"
```
Comment (4brunu) [inline modules/openapi-generator/src/main/resources/kotlin-client/build.gradle.mustache:29]: @kuFEAR I think this line should only be applied when using kotlin serialization, or not?

Comment (kuFEAR) [reply modules/openapi-generator/src/main/resources/kotlin-client/build.gradle.mustache:30]: right, thanks I fixed it

Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/build.gradle.mustache:30]: Thanks

**Comment by 4brunu** [comment]
The PR is quit big, so I could review it the way I wanted, but I don't see anything wrong with it

**Comment by kuFEAR** [comment]
> The PR is quit big, so I could review it the way I wanted, but I don't see anything wrong with it

the major changes is:
- update kotlin and gradle version in samples for ability to use the last kotlinx.serialization version
- switch runs gradle wrapper instead buildin gradle on ci for running gradlew test task on samples, 
- add support to use kotlinx.serialization for jvm/android environment(I didn't change anything in multiplatform, it's better to wait https://github.com/OpenAPITools/openapi-generator/pull/7353 PR)

**Comment by 4brunu** [comment]
Looks good to me 👍

**Comment by kuFEAR** [comment]
@4brunu can you re-run builds on CI? It looks not related with my changes

**Comment by 4brunu** [comment]
I don't have the permissions to do it.
But you can do it, by closing and reopening this PR.

**Comment by kuFEAR** [comment]
@wing328 @jimschubert Hi, can you re-run circleci build? as I notice it failed on downloading fastxml.jackson

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ApiClient.kt.mustache [hunk 2]
```diff
@@ -45,10 +45,16 @@ import com.squareup.moshi.Moshi
 import retrofit2.converter.moshi.MoshiConverterFactory
 {{/moshi}}
 
+{{#kotlinx_serialization}}
+import com.jakewharton.retrofit2.converter.kotlinx.serialization.asConverterFactory
+import org.openapitools.client.infrastructure.Serializer.jvmJson
```
Comment (blendthink) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/jvm-retrofit2/infrastructure/ApiClient.kt.mustache:50]: I suggest.
How about that?

```suggestion
import {{packageName}}.infrastructure.Serializer.jvmJson
```

**Comment by kuFEAR** [comment]
@wing328 Hi! I have PR for some time in repo. Can you approve it? I think it will be really useful for many projects

**Comment by wing328** [comment]
@kuFEAR thanks again for the PR. Let me try to review shortly....

**Comment by wing328** [comment]
@kuFEAR quickly went through it. looks good overall 👍 

Have minor 2 questions. Free in the coming week to discuss it further? PM via Slack to discuss further. Have a nice weekend.

**Comment by wing328** [comment]
As discussed, please recover the following deleted test files:
```
samples/client/petstore/kotlin-string/src/test/kotlin/org/openapitools/client/PetApiTest.kt
samples/client/petstore/kotlin-threetenbp/src/test/kotlin/org/openapitools/client/PetApiTest.kt
samples/client/petstore/kotlin/src/test/kotlin/org/openapitools/client/PetApiTest.kt
samples/client/petstore/kotlin/src/test/kotlin/org/openapitools/client/StoreApiTest.kt
```

**Comment by wing328** [comment]
All tests passed. I may file a PR later to make minor enhancements. 

Thanks again for the PR.

**Comment by wing328** [comment]
@kuFEAR thanks for the PR, which has been included in the v5.1.0 release: https://twitter.com/oas_generator/status/1373636654024380423

**Pull request merged**